active/0000-registry-reform.md

```diff
@@ -0,0 +1,109 @@
+
+# Summary
+
+Fully encapsulate and privatize the `Container` and `Registry` classes by
+exposing a select subset of public methods on `Application` and
+`ApplicationInstance`.
+
+# Motivation
+
+The `Container` and `Registry` classes currently lead a confusing life of
+semi-private exclusion within Ember applications. They are undocumented
+publicly but not fully private either, as knowledge of their particulars is
+required for developing both initializers and unit tests. This situation has
+become untenable as the new `Registry` class has been extracted from
+`Container`, and the complexity of their usage has grown across
+`Application` and `ApplicationInstance` classes.
+
+We can bring sanity to this situation by continuing the work started at the
+`Application` level to expose methods such as `register` and `inject` from the
+internally maintained `Registry`.
+
+Furthermore, once `Container` and `Registry` are fully private, their
+architecture and documentation can be cleaned up. For instance, a
+`Container` can freely reference its associated `Registry` as `registry`
+rather than `_registry`, as it can be assumed that only framework developers
+will reference this property.
+
+# Detailed design
+
+`Application` will expose the following methods from its internally maintained
+registry:
+
+* `register`
+* `inject`
+* `registerOptions` - mapped to `Registry#options`
+* `registerOptionsForType` - mapped to `Registry#optionsForType`
+
+`ApplicationInstance` will also expose the the same methods. However, these
+methods will be exposed from its own internally maintained registry, which
+has the associated `Application`'s registry configured as a "fall back". No
+direct path will be provided from the `ApplicationInstance` to the
+`Application`'s registry.
+
+`ApplicationInstance` will also expose the following methods from its
+internally maintained container:
+
+* `lookup`
+* `lookupFactory`
+
+`ApplicationInstance` will cease exposing `container`, `registry`, and
+`applicationRegistry` publicly.
+
+`Application` initializers will receive a single argument to `initialize`:
+`application`.
+
+Likewise, `ApplicationInstance` initializers will receive a single argument
+to `initialize`: `applicationInstance`.
+
+`Container` and `Registry` will be made fully private and documented as
+such. Each `Container` will freely reference its associated `Registry` as
+`registry` rather than `_registry`.
+
+[ember-test-helpers](https://github.com/switchfly/ember-test-helpers)
+will provide an `isolatedApplicationInstance` method instead of an
+`isolatedContainer` for unit testing. A mechanism will be developed to specify
+which initializers should be engaged in the initialization of this instance.
+In this way, we can avoid duplication of registration logic, as is currently
+done in a most un-DRY manner in the [isolatedContainer](https://github.com/switchfly/ember-test-helpers/blob/master/lib/ember-test-helpers/isolated-container.js#L56-L79).
```

Comment:
For Ember-Data we use `DS._setupContainer`:

  * https://github.com/emberjs/data/blob/87dec3e56b6b1744c49103417cc8cacc8f1edb23/packages/ember-data/lib/setup-container.js#L7

If Ember only needs to register/inject, then I would prefer such a hook since they can be easily layered onto a single registry in an arbitrary order.

However I note that you suggest the name be `isolatedApplicationInstance` - does this imply you will return a real instance and not simply a container with a configured registry?